In [33]:
import polars as pl
import pandas as pd
import numpy as np
import os

DATA_PATH = r"D:\003. HK1 - Năm 3\02. CS116 - Python cho Máy học\CS116-DoAn\dataset\sales_pers.item_chunk_0.parquet"

df = pl.read_parquet(DATA_PATH)
print(f"Số dòng: {df.height:,}, Số cột: {df.width}")


Số dòng: 27,332, Số cột: 34


In [34]:
print("\nCác cột ban đầu:")
print(df.columns)


Các cột ban đầu:
['p_id', 'item_id', 'price', 'category_l1_id', 'category_l1', 'category_l2_id', 'category_l2', 'category_l3_id', 'category_l3', 'category_id', 'category', 'description', 'brand', 'manufacturer', 'creation_timestamp', 'is_deleted', 'created_date', 'updated_date', 'sync_status_id', 'last_sync_date', 'sync_error_message', 'image_url', 'gender_target', 'age_group', 'item_type', 'gp', 'weight', 'color', 'size', 'origin', 'volume', 'material', 'sale_status', 'description_new']


# Task 1: Loại bỏ các cột mà nhóm nghĩ là không cần thiết.

In [35]:
# --- Danh sách cột cần loại bỏ (theo phân tích EDA & tương quan) ---
cols_to_drop = [
    # Cột phân cấp danh mục (trùng lặp với category)
    "category_l1", "category_l2", "category_l3",
    "category_l1_id", "category_l2_id", "category_l3_id",
    "category_id",

    # Cột định lượng thừa / không dùng
    "gp", "weight",

    # Cột metadata hệ thống
    "manufacturer", "is_deleted", "sync_status_id",
    "sync_error_message", "image_url", "last_sync_date",
    "volume", "creation_timestamp", "updated_date", "created_date",

    # Cột ID nội bộ
    "p_id"
]


# --- Loại bỏ các cột không cần thiết ---
df_cleaned = df.drop(cols_to_drop)

print(f"\nĐã loại bỏ {len(cols_to_drop)} cột không cần thiết.")
print(f"Số cột còn lại: {df_cleaned.width}")
print("\nDanh sách cột sau khi loại bỏ:")
print(df_cleaned.columns)

# --- Xuất dữ liệu sau Task 1 ---
OUTPUT_DIR = r"D:\003. HK1 - Năm 3\02. CS116 - Python cho Máy học\CS116-DoAn\data\item-chunk"
os.makedirs(OUTPUT_DIR, exist_ok=True)
OUTPUT_PATH = os.path.join(OUTPUT_DIR, "item_chunk_task1_cleaned.parquet")

df_cleaned.write_parquet(OUTPUT_PATH)

print(f"\nDữ liệu sau Task 1 đã được lưu tại:\n{OUTPUT_PATH}")


Đã loại bỏ 20 cột không cần thiết.
Số cột còn lại: 14

Danh sách cột sau khi loại bỏ:
['item_id', 'price', 'category', 'description', 'brand', 'gender_target', 'age_group', 'item_type', 'color', 'size', 'origin', 'material', 'sale_status', 'description_new']

Dữ liệu sau Task 1 đã được lưu tại:
D:\003. HK1 - Năm 3\02. CS116 - Python cho Máy học\CS116-DoAn\data\item-chunk\item_chunk_task1_cleaned.parquet


In [36]:
df_cleaned

item_id,price,category,description,brand,gender_target,age_group,item_type,color,size,origin,material,sale_status,description_new
str,"decimal[38,4]",str,str,str,str,str,str,str,str,str,str,i32,str
"""0502020000004""",99000.0000,"""Núm ty Dr Brown""","""Không xác định""","""Dr.Brown's""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Hộp 2 núm ty DrBrown's Options Plus cổ rộng Y cu…"
"""0010290040150""",69000.0000,"""Bộ quần áo bé gái""","""Không xác định""","""Con Cưng""","""Bé Gái""","""Từ 3Y""","""Bộ quần áo""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Không xác định"""
"""0008010000015""",45000.0000,"""Gặm nướu khác""","""- Chất liệu: Sản phẩm được làm bằng chất liệu silicone mềm, dẻo và nước đã được chưng cất đảm bảo an…","""Thương hiệu khác""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Miếng Gặm Nướu Papa (CEQ004) (Cá hồng) C…"
"""0020010000094""",401000.0000,"""Merries_Sơ Sinh""","""Tã dán Merries size S 82 miếng là sản phẩm dành cho bé 4-8kg đến từ thương hiệu uy tín Merries của…","""Merries Nhật""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Nhật Bản, Nhật Bản""","""Giấy, bột giấy, vải không dệt, hạt thấm hút, màng polyolefin, cao su tự nhiên, polyurethane, keo dín…",0,"""Không xác định"""
"""0020010000098""",401000.0000,"""Merries_Tã Quần""","""Bỉm tã quần Merries size M 58 miếng là sản phẩm dành cho bé từ 6-11kg đến từ thương hiệu uy tín M…","""Merries Nhật""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Nhật Bản, Nhật Bản""","""Giấy, bột giấy, vải không dệt, hạt thấm hút, màng polyolefin, cao su tự nhiên, polyurethane, keo dín…",0,"""Không xác định"""
"""0024181040235""",49000.0000,"""Áo bé trai""","""Áo thun bé trai tay ngắn CF B078010 Xanh là sản phẩm với chất liệu cotton mềm mại, thấm hút mồ hôi,…","""CF (ConCung Fashion)""","""Bé Trai""","""Không xác định""","""Áo""","""Xanh""","""6-9M""","""Không xác định""","""Không xác định""",0,"""Không xác định"""
"""0008180000017""",425000.0000,"""Thảm xốp mảnh ghép""","""- Sản phẩm dành cho bé từ 3 - 36 tháng tuổi. - Nguyên liệu được làm bằng cao su có tính đàn hồi tốt,…","""Quà tặng không bán""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Thảm xốp hình chữ 30x30cm, 26 miếng Chất…"
"""0020010000150""",355000.0000,"""Huggies Thường_Tã Quần""","""Dòng sản phẩm tã quần Huggies Dry với tinh chất tràm trà tự nhiên sẽ mang đến cho da con cảm giác…","""Huggies Dry""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Việt Nam, Mỹ""","""Vải không dệt,PE, keo dính. hạt thấm hút Polymer, bột giấy, màng polyolefin, polyester, thun đàn hồi…",0,"""Không xác định"""
"""0501030000126""",69000.0000,"""Đầm""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,null


# Task 2: Xử lý NULL, Xử lý Outlier

## Phân tích cơ bản sau khi EDA và tương quan dữ liệu

Trong phần EDA, chúng tôi có thấy rằng description & description_new có độ tương quan khá cao (0.74), và có độ tương quan với các cột: brand, gender_target,age_group, item_type, color, size, origin, material khá cao (từ 0.4 và có thể lên tới 0.78).

Vì vậy, chúng tôi sẽ tiến hành tìm hiểu sâu hơn về 2 cột description và description_new này để xem có thể xử dụng chúng cho một số mục đích khác sau này không.

In [37]:
pl.Config.set_tbl_rows(50)             # số dòng tối đa hiển thị
pl.Config.set_tbl_cols(10)             # số cột tối đa hiển thị
pl.Config.set_tbl_width_chars(200)     # tổng chiều rộng bảng ký tự
pl.Config.set_fmt_str_lengths(2000)     # độ dài chuỗi tối đa hiển thị (quan trọng nhất!)

df_cleaned.select(["description", "description_new"]).head(10)

description,description_new
str,str
"""Không xác định""","""Chi tiết sản phẩm Tên sản phẩm: Hộp 2 núm ty DrBrown's Options Plus cổ rộng Y cut (Trên 9 tháng) Loại sản phẩm: Núm ty Đối tượng sử dụng: Trẻ trên 9 tháng Thương hiệu: DrBrown's Điểm nổi bật: Núm vú silicon siêu mềm giúp phát triển cơ hàm của bé thông qua sự cử động lưỡi tự nhiên. Hộp 2 núm ty DrBrown's Options Plus cổ rộng Y cut (Trên 9 tháng) là núm ty dành cho trẻ trên 9 tháng của thương hiệu DrBrown's đến từ Mỹ. Núm vú silicon siêu mềm giúp phát triển cơ hàm của bé thông qua sự cử động lưỡi một cách tự nhiên. Núm ty silicon mềm mại Núm vú silicon siêu mềm giúp phát triển cơ hàm của bé thông qua sự cử động lưỡi một cách tự nhiên. Thiết kế đầu núm vừa vặn với vòm miệng bé, mang lại sự thoải mái khi sử dụng. Thiết kế thông minh Núm ty DrBrown's Options Plus được thiết kế với lỗ tiết sữa Y cut, phù hợp với nhu cầu ăn cháo đặc của bé trên 9 tháng tuổi, từ đó giúp bé có cảm giác gần gũi hơn trong từng lần bú. Hướng dẫn sử dụng Rửa sạch núm ty bằng nước rửa chuyên dụng trước khi sử dụng lần đầu. Đun sôi núm vú trong vòng 3-5 phút để tiệt trùng sau mỗi lần sử dụng. Kiểm tra núm vú trước khi cho bé sử dụng, đảm bảo không có dấu hiệu rách hoặc thủng. Không cho trẻ bú bình khi không có sự giám sát của người lớn. Hướng dẫn bảo quản Bảo quản ở nhiệt độ phòng, nơi khô ráo, thoáng mát. Tránh ánh nắng mặt trời trực tiếp. Thay núm ty mới sau khi sử dụng từ 1 - 2 tháng để đảm bảo vệ sinh cho trẻ. Lưu ý Không rửa núm ty bằng xà phòng hoặc nước tẩy rửa. Không tiệt trùng núm ty bằng lò vi sóng. Kiểm tra núm vú thường xuyên trước khi sử dụng. Thành phần Silicon cao cấp Không chứa BPA Chất đàn hồi tốt"""
"""Không xác định""","""Không xác định"""
"""- Chất liệu: Sản phẩm được làm bằng chất liệu silicone mềm, dẻo và nước đã được chưng cất đảm bảo an toàn cho bé. - Dành cho các bé trong giai đoạn mọc răng, giúp làm giảm đau, ngứa lợi cho bé. - Bề mặt có các núm mát-xa nhỏ, mềm mịn, tạo cảm giác dễ chịu cho bé. - Có độ đàn hồi tốt, dễ dàng uốn dẻo, bền đẹp không gây ảnh hưởng tới sức khỏe cho bé. - Thiết kế dạng hình các con vật đồ dùng ngộ nghỉnh, đặc biệt có tay cầm, giúp bé dễ cầm nắm. - Trước khi cho bé sử dụng, bạn có thể cho vào tủ lạnh tạo cảm giác mát lạnh bé thích thú Lưu ý: Không sử dụng nước sôi để khử trùng sản phẩm. Xuất xứ: Thái Lan""","""Chi tiết sản phẩm Tên sản phẩm: Miếng Gặm Nướu Papa (CEQ004) (Cá hồng) Chất liệu: Silicone mềm, dẻo và nước đã được chưng cất Đối tượng sử dụng: Dành cho các bé trong giai đoạn mọc răng Thương hiệu: Papa Xuất xứ: Thái Lan Giá: 45.000₫ **Miếng Gặm Nướu Papa (CEQ004) (Cá hồng)** là một sản phẩm giúp bé giảm đau trong giai đoạn mọc răng của thương hiệu **Papa đến từ Thái Lan**. Sản phẩm được thiết kế đặc biệt để tạo cảm giác dễ chịu cho bé khi nhai. Chất liệu an toàn: Silicone mềm, dẻo Miếng gặm nướu được làm bằng chất liệu silicone mềm, dẻo và nước đã được chưng cất, đảm bảo an toàn tuyệt đối cho sức khỏe của bé. Cảm giác dễ chịu: Bề mặt núm mát-xa Bề mặt có các núm mát-xa nhỏ, mềm mịn, tạo cảm giác dễ chịu và thích thú khi bé sử dụng. Thiết kế ngộ nghĩnh: Hình động vật với tay cầm Sản phẩm được thiết kế dạng hình các con vật đồ dùng ngộ nghĩnh, đặc biệt có tay cầm giúp bé dễ cầm nắm và chơi đùa. Độ đàn hồi tốt: Bền đẹp và dễ sử dụng Miếng gặm nướu có độ đàn hồi tốt, dễ dàng uốn dẻo, bền đẹp mà không gây ảnh hưởng đến sức khỏe cho bé. Cảm giác mát lạnh: Hỗ trợ giảm đau Trước khi cho bé sử dụng, bạn có thể cho vào tủ lạnh tạo cảm giác mát lạnh, giúp bé thích thú và giảm đau nướu. Hướng dẫn sử dụng Rửa sạch sản phẩm trước khi cho bé sử dụng. Để sản phẩm trong tủ lạnh trước khi cho bé sử dụng để tạo cảm giác mát lạnh. Giám sát bé khi sử dụng để đảm bảo an toàn. Hướng dẫn bảo quản Bảo quản sản phẩm ở nơi khô ráo, thoáng mát. Không sử dụng nước sôi để khử trùng sản phẩm. Thường xuyên kiểm tra sản phẩm để đảm bảo không bị hư hỏng. Lưu ý Sản phẩm chỉ dành cho bé trong giai đoạn mọc răng. Khi sản phẩm có dấu hiệu hư hỏng, không nên cho bé sử dụng. …"

In [38]:
kxd_list = ["không xác định","null"]

s_new = pl.col("description_new").cast(pl.Utf8).str.to_lowercase().str.strip_chars()
s_old = pl.col("description").cast(pl.Utf8).str.to_lowercase().str.strip_chars()

new_has_text = pl.col("description_new").is_not_null() & (s_new.str.len_chars() > 0) & (~s_new.is_in(kxd_list))
old_has_text = pl.col("description").is_not_null() & (s_old.str.len_chars() > 0) & (~s_old.is_in(kxd_list))

counts_value = (
    df_cleaned.select(
        (new_has_text & old_has_text).sum().alias("both_present"),
        (new_has_text & ~old_has_text).sum().alias("only_description_new"),
        (~new_has_text & old_has_text).sum().alias("only_description"),
        (~new_has_text & ~old_has_text).sum().alias("neither_present"),
    )
    .with_columns(
        total = pl.sum_horizontal(pl.all())
    )
    .with_columns(
        *[(pl.col(c) / pl.col("total") * 100).round(2).alias(f"{c}_pct")
          for c in ["both_present","only_description_new","only_description","neither_present"]]
    )
)
print(counts_value)

shape: (1, 9)
┌──────────────┬──────────────────────┬──────────────────┬─────────────────┬───────┬──────────────────┬──────────────────────────┬──────────────────────┬─────────────────────┐
│ both_present ┆ only_description_new ┆ only_description ┆ neither_present ┆ total ┆ both_present_pct ┆ only_description_new_pct ┆ only_description_pct ┆ neither_present_pct │
│ ---          ┆ ---                  ┆ ---              ┆ ---             ┆ ---   ┆ ---              ┆ ---                      ┆ ---                  ┆ ---                 │
│ u32          ┆ u32                  ┆ u32              ┆ u32             ┆ u32   ┆ f64              ┆ f64                      ┆ f64                  ┆ f64                 │
╞══════════════╪══════════════════════╪══════════════════╪═════════════════╪═══════╪══════════════════╪══════════════════════════╪══════════════════════╪═════════════════════╡
│ 6308         ┆ 3194                 ┆ 2395             ┆ 15435           ┆ 27332 ┆ 23.08            ┆ 11

**Độ phủ hai cột mô tả: `description_new` vs `description`**

**Bảng tóm tắt thông tin:**

| Trường hợp | Số dòng | Tỷ lệ (%) |
|---|---:|---:|
| **Cả hai đều có** (`both_present`) | **6,308** | **23.08** |
| **Chỉ có `description_new`** | **3,194** | **11.69** |
| **Chỉ có `description`** | **2,395** | **8.76** |
| **Không có cả hai** (`neither_present`) | **15,435** | **56.47** |
| **Tổng** | **27,332** | **100.00** |

- Hàng **có `description_new`** = 6,308 + 3,194 = **9,502** (**34.77%**).  
  ⟶ **Thiếu `description_new`** ≈ **17,830**.
- Hàng **có `description`** = 6,308 + 2,395 = **8,703** (**31.84%**).  
  ⟶ **Thiếu `description`** ≈ **18,629**.
- Hàng **có ít nhất một** mô tả = 27,332 − 15,435 = **11,897** (**43.53%**).  


### ⟶ **Dự định tiếp theo**: Gộp description và description_new lại để bổ sung dữ liệu cho nhau, giảm chiều dữ liệu



#### **BÀI TOÁN MỚI CẦN GIẢI QUYẾT**: Nên gộp `description` vào `description_new` hay ngược lại?  


In [39]:
import polars as pl
import re

# ===== 1) Cấu hình =====
LEVEL = "word"   # "word" hoặc "char"
N = 2            # n-gram: 1/2/3... ; ví dụ word-bigram = 2

# Các giá trị được coi là "thiếu"
PLACEHOLDERS = [
    "", "-", "--",
    "không xác định", "khong xac dinh", "không xac dinh",
    "không rõ", "chưa rõ", "không có thông tin",
    "n/a", "na", "none", "null", "unk", "unknown"
]

# ===== 2) Hàm tiện ích =====
def tokenize(s: str):
    return re.findall(r"\w+", s.lower())

def ngrams(tokens, n=2):
    if len(tokens) < n: 
        return []
    return [" ".join(tokens[i:i+n]) for i in range(len(tokens)-n+1)]

def build_set(s: str, level="word", n=2):
    if not s:
        return set()
    if level == "char":
        s_norm = re.sub(r"\s+", " ", s.lower()).strip()
        if len(s_norm) < n:
            return set()
        return set(s_norm[i:i+n] for i in range(len(s_norm)-n+1))
    else:
        toks = tokenize(s)
        return set(toks) if n == 1 else set(ngrams(toks, n))

def contain_ratio_sets(A_set: set, B_set: set) -> float:
    return 0.0 if not A_set else len(A_set & B_set) / len(A_set)

def contain_pair(a: str, b: str, level="word", n=2):
    A = build_set(a, level=level, n=n)
    B = build_set(b, level=level, n=n)
    r = contain_ratio_sets(A, B)  # Contain(A|B)
    p = contain_ratio_sets(B, A)  # Contain(B|A)
    f1 = 0.0 if (p + r) == 0 else 2 * p * r / (p + r)
    return (r, p, f1, len(A), len(B))

# Chuẩn hoá text để so đối chiếu placeholder
def norm_expr(colname: str):
    return (
        pl.col(colname).cast(pl.Utf8)
        .str.to_lowercase()
        .str.replace_all(r"\s+", " ")
        .str.strip_chars(" .,-:;—–_/\\|\"'`")  # bỏ dấu câu ở đầu/cuối
        .str.strip_chars()                     # bỏ khoảng trắng thừa
    )

def is_missing(colname: str):
    n = norm_expr(colname)
    return pl.col(colname).is_null() | n.is_in(PLACEHOLDERS)

# ===== 3) Mask both_present với định nghĩa "thiếu" mới =====
miss_desc     = is_missing("description")
miss_desc_new = is_missing("description_new")
both_present  = (~miss_desc) & (~miss_desc_new)

# Báo cáo thiếu dữ liệu (fraction)
null_report = df_cleaned.select([
    (miss_desc & miss_desc_new).mean().alias("both_null_frac"),
    (miss_desc & (~miss_desc_new)).mean().alias("desc_null_only_frac"),
    ((~miss_desc) & miss_desc_new).mean().alias("desc_new_null_only_frac"),
    both_present.mean().alias("both_present_frac"),
])
print("=== Null report (fraction) — placeholders treated as null ===")
print(null_report)

# ===== 4) Tính containment chỉ trên both_present =====
both_df = df_cleaned.filter(both_present)

res = both_df.with_columns([
    pl.struct(["description", "description_new"]).map_elements(
        lambda row: contain_pair(row["description"], row["description_new"], level=LEVEL, n=N)[0]
    ).alias("contain_desc_in_descnew"),   # recall: A|B

    pl.struct(["description", "description_new"]).map_elements(
        lambda row: contain_pair(row["description"], row["description_new"], level=LEVEL, n=N)[1]
    ).alias("contain_descnew_in_desc"),   # precision: B|A

    pl.struct(["description", "description_new"]).map_elements(
        lambda row: contain_pair(row["description"], row["description_new"], level=LEVEL, n=N)[2]
    ).alias("contain_f1"),
])

summary = res.select([
    pl.len().alias("n_both_present"),
    pl.col("contain_desc_in_descnew").mean().alias("mean_contain(A|B)"),
    pl.col("contain_descnew_in_desc").mean().alias("mean_contain(B|A)"),
    pl.col("contain_f1").mean().alias("mean_contain_F1"),
    pl.col("contain_desc_in_descnew").quantile(0.5).alias("median_contain(A|B)"),
    pl.col("contain_desc_in_descnew").quantile(0.9).alias("p90_contain(A|B)"),
])
print(f"=== Containment summary (LEVEL={LEVEL}, N={N}) — after placeholder filtering ===")
print(summary)

# (Tuỳ chọn) kiểm tra vài dòng F1 cao
print(res.select(["description", "description_new", "contain_desc_in_descnew", "contain_descnew_in_desc", "contain_f1"])
        .sort("contain_f1", descending=True)
        .head(10))


=== Null report (fraction) — placeholders treated as null ===
shape: (1, 4)
┌────────────────┬─────────────────────┬─────────────────────────┬───────────────────┐
│ both_null_frac ┆ desc_null_only_frac ┆ desc_new_null_only_frac ┆ both_present_frac │
│ ---            ┆ ---                 ┆ ---                     ┆ ---               │
│ f64            ┆ f64                 ┆ f64                     ┆ f64               │
╞════════════════╪═════════════════════╪═════════════════════════╪═══════════════════╡
│ 0.564723       ┆ 0.116859            ┆ 0.087626                ┆ 0.230792          │
└────────────────┴─────────────────────┴─────────────────────────┴───────────────────┘
=== Containment summary (LEVEL=word, N=2) — after placeholder filtering ===
shape: (1, 6)
┌────────────────┬───────────────────┬───────────────────┬─────────────────┬─────────────────────┬──────────────────┐
│ n_both_present ┆ mean_contain(A|B) ┆ mean_contain(B|A) ┆ mean_contain_F1 ┆ median_contain(A|B) ┆ p90_cont

**mean Contain(A|B) = 0.623**
- Trung bình 62.3% bigram trong description cũng có trong description_new.
- Khi có dữ liệu thật, description_new bao phủ phần lớn nội dung của description.

**mean Contain(B|A) = 0.324**
- Chỉ 32.4% bigram trong description_new xuất hiện trong description.
-  description_new giàu/chi tiết hơn, có nhiều cụm mới (vd. “Tên sản phẩm/Thương hiệu/Xuất xứ/Kích thước…”) mà description không có.
-  Đây là lý do precision thấp.

**mean F1 = 0.405**
- Trung bình độ khớp hai chiều chỉ ở mức vừa (hàm hòa hợp của hai contain). Bị kéo xuống bởi B|A thấp.

**median Contain(A|B) = 0.620; p90 = 0.807**
- 50% cặp có ≥62% nội dung description được bao bởi description_new; top 10% đạt ≥80.7%.
- Phân bố khá lệch: nhiều cặp khớp tốt, nhưng cũng có không ít cặp lệch mạnh (kéo mean xuống).

### Xử lý  description và description_new

### Gộp 2 cột description và description_new

- Sau khi đã xác định các giá trị thiếu chỉ gồm null hoặc "Không xác định" thì chỉ cần kiểm tra missing(x) = is_null(x) OR lower(strip(x)) == "Không xác định" 
- Để nhận diện thiếu, tránh copy nhầm "Không xác định"

**Logic gộp**
1. Nếu des_new không missing => lấy des_new
2. Nếu des_new missing nhưng des không missing => lấy des 
3. Nếu cả hai missing => Để NULL
4. Thêm cột description_source = new / old_as_fill / missing để audit

In [40]:
pl.Config.set_tbl_rows(20) 
pl.Config.set_tbl_cols(20) 
pl.Config.set_tbl_width_chars(150)
pl.Config.set_fmt_str_lengths(100)

polars.config.Config

In [41]:
# --- 1️⃣ Định nghĩa hàm phát hiện "thiếu dữ liệu" ---
def is_missing(col: str) -> pl.Expr:
    c = pl.col(col)
    return c.is_null() | (
        c.cast(pl.Utf8)
         .str.to_lowercase()
         .str.strip_chars()
         .is_in(["không xác định", "null", "none", ""])
    )

# --- 2️⃣ Tạo mask ---
miss_new = is_missing("description_new")
miss_old = is_missing("description")

# --- 3️⃣ Gộp hai mô tả và đánh dấu nguồn ---
df_cleaned = df_cleaned.with_columns([
    pl.when(~miss_new).then(pl.col("description_new"))
      .when(~miss_old).then(pl.col("description"))
      .otherwise(None)
      .alias("description_merge"),

    pl.when(~miss_new).then(pl.lit("from_new"))
     .when(~miss_old).then(pl.lit("from_old"))
     .otherwise(pl.lit("missing"))
     .alias("description_source")
])

# --- 4️⃣ Báo cáo tỷ lệ merge ---
merge_report = (
    df_cleaned.select(
        pl.count().alias("total"),
        (pl.col("description_source") == "from_new").sum().alias("from_new"),
        (pl.col("description_source") == "from_old").sum().alias("filled_from_old"),
        (pl.col("description_source") == "missing").sum().alias("still_missing"),
    )
    .with_columns([
        (pl.col("from_new") / pl.col("total") * 100).round(2).alias("from_new_pct"),
        (pl.col("filled_from_old") / pl.col("total") * 100).round(2).alias("filled_from_old_pct"),
        (pl.col("still_missing") / pl.col("total") * 100).round(2).alias("still_missing_pct"),
    ])
)

print("=== 🧩 Merge report (description + description_new) ===")
print(merge_report)

# --- 5️⃣ Xem vài dòng mẫu được fill từ mô tả cũ ---
print("\n=== 🔍 Sample rows (filled_from_old) ===")
print(
    df_cleaned.filter(pl.col("description_source") == "from_old")
              .select(["description_new", "description", "description_merge"])
              .head(5)
)



=== 🧩 Merge report (description + description_new) ===
shape: (1, 7)
┌───────┬──────────┬─────────────────┬───────────────┬──────────────┬─────────────────────┬───────────────────┐
│ total ┆ from_new ┆ filled_from_old ┆ still_missing ┆ from_new_pct ┆ filled_from_old_pct ┆ still_missing_pct │
│ ---   ┆ ---      ┆ ---             ┆ ---           ┆ ---          ┆ ---                 ┆ ---               │
│ u32   ┆ u32      ┆ u32             ┆ u32           ┆ f64          ┆ f64                 ┆ f64               │
╞═══════╪══════════╪═════════════════╪═══════════════╪══════════════╪═════════════════════╪═══════════════════╡
│ 27332 ┆ 9502     ┆ 2395            ┆ 15435         ┆ 34.77        ┆ 8.76                ┆ 56.47             │
└───────┴──────────┴─────────────────┴───────────────┴──────────────┴─────────────────────┴───────────────────┘

=== 🔍 Sample rows (filled_from_old) ===
shape: (5, 3)
┌─────────────────┬─────────────────────────────────────────────────────────────────┬───────

C:\Users\PC\AppData\Local\Temp\ipykernel_12644\9102509.py:31: DeprecationWarning: `pl.count()` is deprecated. Please use `pl.len()` instead.
(Deprecated in version 0.20.5)
  pl.count().alias("total"),


In [42]:
df_cleaned.head()

item_id,price,category,description,brand,gender_target,age_group,item_type,color,size,origin,material,sale_status,description_new,description_merge,description_source
str,"decimal[38,4]",str,str,str,str,str,str,str,str,str,str,i32,str,str,str
"""0502020000004""",99000.0000,"""Núm ty Dr Brown""","""Không xác định""","""Dr.Brown's""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Hộp 2 núm ty DrBrown's Options Plus cổ rộng Y cu…","""Chi tiết sản phẩm Tên sản phẩm: Hộp 2 núm ty DrBrown's Options Plus cổ rộng Y cu…","""from_new"""
"""0010290040150""",69000.0000,"""Bộ quần áo bé gái""","""Không xác định""","""Con Cưng""","""Bé Gái""","""Từ 3Y""","""Bộ quần áo""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Không xác định""",null,"""missing"""
"""0008010000015""",45000.0000,"""Gặm nướu khác""","""- Chất liệu: Sản phẩm được làm bằng chất liệu silicone mềm, dẻo và nước đã được chưng cất đảm bảo an…","""Thương hiệu khác""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Miếng Gặm Nướu Papa (CEQ004) (Cá hồng) C…","""Chi tiết sản phẩm Tên sản phẩm: Miếng Gặm Nướu Papa (CEQ004) (Cá hồng) C…","""from_new"""
"""0020010000094""",401000.0000,"""Merries_Sơ Sinh""","""Tã dán Merries size S 82 miếng là sản phẩm dành cho bé 4-8kg đến từ thương hiệu uy tín Merries của…","""Merries Nhật""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Nhật Bản, Nhật Bản""","""Giấy, bột giấy, vải không dệt, hạt thấm hút, màng polyolefin, cao su tự nhiên, polyurethane, keo dín…",0,"""Không xác định""","""Tã dán Merries size S 82 miếng là sản phẩm dành cho bé 4-8kg đến từ thương hiệu uy tín Merries của…","""from_old"""
"""0020010000098""",401000.0000,"""Merries_Tã Quần""","""Bỉm tã quần Merries size M 58 miếng là sản phẩm dành cho bé từ 6-11kg đến từ thương hiệu uy tín M…","""Merries Nhật""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Nhật Bản, Nhật Bản""","""Giấy, bột giấy, vải không dệt, hạt thấm hút, màng polyolefin, cao su tự nhiên, polyurethane, keo dín…",0,"""Không xác định""","""Bỉm tã quần Merries size M 58 miếng là sản phẩm dành cho bé từ 6-11kg đến từ thương hiệu uy tín M…","""from_old"""


Sau khi đã kiểm tra xong và thấy mọi thứ đã hợp lý, tôi sẽ drop các cột cũ không cần dùng đến (description, description_new, description_source), chỉ giữ lại cột description_merge

In [ ]:
# --- Xóa cột cũ để giữ dataset gọn gàng ---
df_cleaned = df_cleaned.drop(["description", "description_new", "description_source"])

In [44]:
df_cleaned.shape
df_cleaned.head(5)

item_id,price,category,brand,gender_target,age_group,item_type,color,size,origin,material,sale_status,description_merge
str,"decimal[38,4]",str,str,str,str,str,str,str,str,str,i32,str
"""0502020000004""",99000.0000,"""Núm ty Dr Brown""","""Dr.Brown's""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Hộp 2 núm ty DrBrown's Options Plus cổ rộng Y cu…"
"""0010290040150""",69000.0000,"""Bộ quần áo bé gái""","""Con Cưng""","""Bé Gái""","""Từ 3Y""","""Bộ quần áo""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,null
"""0008010000015""",45000.0000,"""Gặm nướu khác""","""Thương hiệu khác""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""",0,"""Chi tiết sản phẩm Tên sản phẩm: Miếng Gặm Nướu Papa (CEQ004) (Cá hồng) C…"
"""0020010000094""",401000.0000,"""Merries_Sơ Sinh""","""Merries Nhật""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Nhật Bản, Nhật Bản""","""Giấy, bột giấy, vải không dệt, hạt thấm hút, màng polyolefin, cao su tự nhiên, polyurethane, keo dín…",0,"""Tã dán Merries size S 82 miếng là sản phẩm dành cho bé 4-8kg đến từ thương hiệu uy tín Merries của…"
"""0020010000098""",401000.0000,"""Merries_Tã Quần""","""Merries Nhật""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Không xác định""","""Nhật Bản, Nhật Bản""","""Giấy, bột giấy, vải không dệt, hạt thấm hút, màng polyolefin, cao su tự nhiên, polyurethane, keo dín…",0,"""Bỉm tã quần Merries size M 58 miếng là sản phẩm dành cho bé từ 6-11kg đến từ thương hiệu uy tín M…"


## Xử lý NULL, Outlier sau khi phân tích

# Task 3: Phân tích tương đồng và xác định xem các thuộc tính tương tự nhau. Từ đó loại bỏ đặc trưng thừa.

# Task 4: Chuẩn hóa dữ liệu (nếu có), biến đổi dữ liệu

# Task 5: Nhóm hãy suy nghĩ xem, với bài toán dự đoán mua hàng, ta có thể tạo mới những đặc trưng nào. Sau đó tiến hành rút trích thêm các đặc trưng. Task này rất quan trọng vì ảnh hưởng hiệu quả của hệ thống.

